### DISCUSSION TITLE
Utlizing Compressed Trajectory format for Swarming?

### DISCUSSION THREAD

**Discussion by csanchez9427** [bitcraze/discussions#837] [Q&A]
I currently use the crazyswarm platform for basic flight patterns with multiple crazyflies. However, the lab i research at currently uses MATLAB to calculate Bernstein polynomials for all of the trajectory generation. Our current method to implement these trajectories is to calculate the position and velocity throughout the flight path and send them to the crazyswarm server using a ROS at a rate of 10-30Hz. However, we plan to expand to a much larger swarm and are worried about bandwidth.

We wanted to look into the compressed trajectory format, which seems to accept just the control points of the polynomial and the position, velocity, etc. is calculated onboard. This functionality is not yet integrated into crazyswarm or crazyswarm2. I've found a few examples from cflib utilizing the compressed trajectory format, but these examples are only made for flying a single drone and the swarming examples I've looked at don't seem to be able to change paths in real time.

I was wondering if it was possible to use cflib to use the compressed trajectory format to control multiple drones and if it is possible to send control points and have the drone change its trajectory in real time?  I want to be able to just send control points once every couple of seconds instead of sending commands to all of the drones many times per second.

**Comment by krichardsson** [comment]
The trajectory functionality (non-compressed) was added by the Crazyswarm team and I assume there is support for it in Crazyswarm, but I'm not sure. The compressed format was added later by [Collmot for SkyBrush](https://skybrush.io/) to be able to store longer trajectories in RAM, and I can imagine that this is not implemented in Crazyswarm. Both formats provide the same functionality, the only difference is how much RAM that is required.

> I was wondering if it was possible to use cflib to use the compressed trajectory format to control multiple drones

Yes, that should be possible. 

> and if it is possible to send control points and have the drone change its trajectory in real time?

Yes and no. The trajectory concept is based on the idea that you first upload a trajectory and then tell the Crazyflie to execute it, start to end. It is possible to upload multiple trajectories and this is one possible solution to your problem. You can upload a first trajectory and tell the Crayflie to execute it. While this is done, you can upload a second trajectory and when the first is finished, you can tell the Crazyflie to execute the next one, and so on. The problem might be to kick off the next trajectory at the right time, there is not automatic functionality for this. You must also make sure that the position, velocity, acceleration matches when going from one trajectory to the next one for a smooth flight.

If you plan to send new, short trajectories to the Crazyflies fairly often you can probably use the non-compressed format as there would be enough RAM to handle that.

There is also a goto functionality in the high level commander that you can use. You simply tell the Crazyflie to goto a spceific coordinate, at this point the high level commander will plan a new trajectory from the current **state** to go to the desired coordinate and hoover there. The key here is that it will re-plan from the current state, that is you don't have to care about what the current velocity or acceleration is, the planer will do that for you. A trick that has been used in the past is to continuously feed new gotos (before it reaches the target position) to the high level commander re-plan the trajectory an thus get a smooth flight. It will not provide the same degree of control, but might be an option.
